### PR TITLE
ceph_volume_client: allow volumes without namespace isolation

### PR DESCRIPTION
**Pull request by ajarr**
Fixes: https://tracker.ceph.com/issues/23695

Signed-off-by: Ramana Raja <rraja@redhat.com>

### CODE REVIEW

**Comment by ajarr** [comment]
With this patch, if you authorize an ID, client.john,   'read-write' access to a volume1, and then authorize client.john,'read' access to volume2 that belongs to the same data pool. client.john's OSD caps of data pool would be set to 'read'. The latest changes to OSD data pool caps takes effect. I think the OSD data pool caps should pick the more permissive caps, here 'read-write'.

@batrick , what do you think?

**Comment by batrick** [comment]
Use case for posterity: https://github.com/kubernetes-incubator/external-storage/pull/674

**Comment by batrick** [comment]
> With this patch, if you authorize an ID, client.john, 'read-write' access to a volume1, and then authorize client.john,'read' access to volume2 that belongs to the same data pool. client.john's OSD caps of data pool would be set to 'read'. The latest changes to OSD data pool caps takes effect.

Ah, I think this is what you were trying to tell me in standup (sorry, my brain doesn't work that well in the morning!).

>  I think the OSD data pool caps should pick the more permissive caps, here 'read-write'.

Given the circumstances, I agree that this is the best option.

**Comment by batrick** [comment]
@ajarr this will need a QA test too.

File: src/pybind/ceph_volume_client.py [hunk 1]
```diff
@@ -633,10 +634,11 @@ def create_volume(self, volume_path, size=None, data_isolated=False):
...
 
-        # enforce security isolation, use seperate namespace for this volume
-        namespace = "{0}{1}".format(self.pool_ns_prefix, volume_path.volume_id)
-        log.info("create_volume: {0}, using rados namespace {1} to isolate data.".format(volume_path, namespace))
-        self.fs.setxattr(path, 'ceph.dir.layout.pool_namespace', namespace, 0)
+        # enforce security isolation, use separate namespace for this volume
+        if namespace_isolated:
+            namespace = "{0}{1}".format(self.pool_ns_prefix, volume_path.volume_id)
+            log.info("create_volume: {0}, using rados namespace {1} to isolate data.".format(volume_path, namespace))
+            self.fs.setxattr(path, 'ceph.dir.layout.pool_namespace', namespace, 0)
```
Comment (cofyc) [inline src/pybind/ceph_volume_client.py]: @ajarr If namespace is isolated, we set `ceph.dir.layout.pool_namespace` on path, which will also set `ceph.dir.layout.pool` (inherited from ancestor) on it. If namespace is not isolated, should we configure `ceph.dir.layout.pool` on volume path intentionally? I'm afraid if ancestor pool changed, we may have permission issues.

File: src/pybind/ceph_volume_client.py [hunk 2]
```diff
@@ -1187,15 +1201,17 @@ def _deauthorize(self, volume_path, auth_id):
...
+            namespace = None
 
         # The auth_id might have read-only or read-write mount access for the
         # volume path.
         access_levels = ('r', 'rw')
         want_mds_caps = {'allow {0} path={1}'.format(access_level, path)
                          for access_level in access_levels}
-        want_osd_caps = {'allow {0} pool={1} namespace={2}'.format(
-                         access_level, pool_name, namespace)
+        want_osd_caps = {'allow {0} pool={1} namespace={2}'.format(access_level, pool_name, namespace)
```
Comment (cofyc) [inline src/pybind/ceph_volume_client.py]: Should we remove `namespace` if namespace is not set?

**Comment by ajarr** [comment]
> With this patch, if you authorize an ID, client.john, 'read-write' access to a volume1, and then authorize client.john,'read' access to volume2 that belongs to the same data pool. client.john's OSD caps of data pool would be set to 'read'. The latest changes to OSD data pool caps takes effect. I think the OSD data pool caps should pick the more permissive caps, here 'read-write'.

Implemented this.

**Comment by ajarr** [comment]
> this will need a QA test too.

I will add that tomorrow.

File: src/pybind/ceph_volume_client.py (hunk 1 above)
Comment (ajarr) [reply src/pybind/ceph_volume_client.py]: @cofyc , I looked at https://github.com/kubernetes-incubator/external-storage/pull/674/commits/29131951c375a8f9c8d4a03e3ebbb3a485e67175
makes sense. thanks!

Comment (ajarr) [reply src/pybind/ceph_volume_client.py]: @cofyc done.

**Comment by ajarr** [comment]
> this will need a QA test too.

done

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-07_15:40:31-fs-wip-pdonnell-testing-20180507.140035-testing-basic-smithi/

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-07_18:27:34-fs-wip-pdonnell-testing-20180507.140035-testing-basic-smithi/

**Comment by batrick** [comment]
Let me know when you're ready to have this go through another round of QA @ajarr

**Comment by ajarr** [comment]
@batrick it's ready

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-08_22:01:43-fs-wip-pdonnell-testing-20180508.195723-testing-basic-smithi/

**Comment by cofyc** [comment]
@batrick @ajarr hi, will this be cherry-picked into luminos branch?

**Comment by batrick** [comment]
@cofyc see https://tracker.ceph.com/issues/24055

**Comment by cofyc** [comment]
Thanks!

**Pull request merged**